### PR TITLE
vk-to-ascii

### PR DESCRIPTION
**Pull request by patrick-lunarg**
This PR sets up Vulkan support for gfxrecon-toascii.

Example output:
```
"[60]vkCreateBuffer": {
    "return": "VK_SUCCESS",
    "device": "0x0000000000000004",
    "pCreateInfo": {
        "sType": "VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO",
        "pNext": null,
        "flags": 0,
        "size": 262144,
        "usage": 1,
        "sharingMode": "VK_SHARING_MODE_EXCLUSIVE",
        "queueFamilyIndexCount": 0,
        "pQueueFamilyIndices": []
    },
    "pAllocator": null,
    "[out]pBuffer": "0x0000000000000025"
},
"[61]vkGetBufferMemoryRequirements": {
    "device": "0x0000000000000004",
    "buffer": "0x0000000000000025",
    "[out]pMemoryRequirements": {
        "size": 262144,
        "alignment": 64,
        "memoryTypeBits": 7
    }
},
"[62]vkAllocateMemory": {
    "return": "VK_SUCCESS",
    "device": "0x0000000000000004",
    "pAllocateInfo": {
        "sType": "VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO",
        "pNext": null,
        "allocationSize": 262144,
        "memoryTypeIndex": 0
    },
    "pAllocator": null,
    "[out]pMemory": "0x0000000000000026"
},
"[63]vkBindBufferMemory": {
    "return": "VK_SUCCESS",
    "device": "0x0000000000000004",
    "buffer": "0x0000000000000025",
    "memory": "0x0000000000000026",
    "memoryOffset": 0
},
"[64]vkMapMemory": {
    "return": "VK_SUCCESS",
    "device": "0x0000000000000004",
    "memory": "0x0000000000000026",
    "offset": 0,
    "size": 18446744073709551615,
    "flags": 0,
    "[out]ppData": "0x00000051A851F1B0"
},
"[65]vkUnmapMemory": {
    "device": "0x0000000000000004",
    "memory": "0x0000000000000026"
}
```

### CODE REVIEW

File: framework/decode/custom_vulkan_to_string.cpp [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+    );
+}
+
+template <>
+std::string ToString<VkAccelerationStructureGeometryDataKHR>(const VkAccelerationStructureGeometryDataKHR& obj, ToStringFlags toStringFlags, uint32_t tabCount, uint32_t tabSize)
+{
+    return ObjectToString(toStringFlags, tabCount, tabSize,
+        [&](std::stringstream& strStrm)
+        {
+            FieldToString(strStrm, true, "triangles", toStringFlags, tabCount, tabSize, ToString(obj.triangles, toStringFlags, tabCount, tabSize));
```
Comment (bradgrantham-lunarg) [inline framework/decode/custom_vulkan_to_string.cpp:72]: To add to Locke's comment from the previous PR - My preference in general would be to pick the right union member keyed from the parent class, where possible.  Like in this case, the acceleration structure input is either triangles, aabbs (for procedural raytracing shaders) or instances (for the top-level acceleration structure build).  For the ASCII output to be valuable to any kind of visual inspection, I would personally like the tool to parse the *correct* union member for me.  The others will be nonsensical and will make it difficult for a human to read the output.

**Review by bradgrantham-lunarg — CHANGES_REQUESTED** [review]

**Comment by bradgrantham-lunarg** [comment]
Hi!  It's so nice to see a Vulkan capture in structured output!

I think there still may be some missing quotes in some methods, though.  I wrote this python:
```
import sys
import json

input = sys.stdin.read()
doc = json.loads(input)
print(doc)
```
but received the following output:
```
Traceback (most recent call last):
  File "parseit.py", line 5, in <module>
    doc = json.loads(input)
  File "C:\Users\grantham\AppData\Local\Programs\Python\Python38\lib\json\__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "C:\Users\grantham\AppData\Local\Programs\Python\Python38\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Users\grantham\AppData\Local\Programs\Python\Python38\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 64 column 23 (char 1787)
```
that line is ```        "deviceType": VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU,```, and I think the value not having quotes around it may be the issue.

FYI this was from a capture of vkcube.  For reference, the binary capture is on the internal file share at https://drive.google.com/file/d/1HC_-S0lBXcWRIthU-wpZWLmRPe0cRlw_/view?usp=sharing

Thanks!

File: framework/decode/custom_vulkan_to_string.cpp (hunk 1 above)
Comment (patrick-lunarg) [reply framework/decode/custom_vulkan_to_string.cpp:96]: Done

Comment (bradgrantham-lunarg) [reply framework/decode/custom_vulkan_to_string.cpp:96]: Can this function be deleted, then?

**Review by bradgrantham-lunarg — APPROVED** [review]

**Review by locke-lunarg — APPROVED** [review]

**Pull request merged**